### PR TITLE
feat(ContextSelector): Add PF4 ContextSelector Component

### PR DESCRIPTION
**Pull request by tlabaj**
#1293, #1525

**What**:

**Additional issues**:

### CODE REVIEW

**Comment by jgiardino** [comment]
One thing I noticed is that during a search, if I enter a lowercase "a", it returns nothing. Should the search be case-insensitive for this component?

The component isn't keyboard accessible at the moment. For the first implementation, I think the simplest solution is to handle the panel similar to how we handle the popover or modal with regard to trapping and managing focus.

- Given the toggle has focus, when the user hits Enter (or the user clicks the toggle), then:
  - The panel `.pf-c-context-selector__menu` traps focus - This behavior was added to the modal in https://github.com/patternfly/patternfly-react/pull/1011. I'm not sure if the popover uses a similar or different method, and if so, which one would make the most sense for the dropdown. The popover might make the most sense given that it also requires the ability to handle positioning relative to the toggle.
  - The first element in the panel that can receive focus has focus. In this case it would be the search field.
- Given the panel is open, 
  - When the user hits Tab, then the next element that can receive focus gets focus (and Shift-Tab goes to the previous item that can receive focus)
  - And given the last item has focus, when the user hits Tab...?
    - In the modal or popover, we would shift focus to the first element that can receive focus. I think for now, that might be the simplest thing to do here. We can always refine this behavior once we have something that works across all experiences (mouse, keyboard, screen reader), but I would recommend handling this as a separate PR given the amount of effort involved.

**Comment by tlabaj** [comment]
@Jenn, the application implements the search functionality.  I can change the example to be case insensitive if you would like.

**Comment by jgiardino** [comment]
Thanks, @tlabaj!

> @jgiardino the application implements the search functionality. I can change the example to be case insensitive if you would like.

I'm not sure what the general expectation is here. That makes sense that we would be more hands-off and rely on consumers to handle that functionality. But I could also see the case where consumers would want to see an example that shows how we would recommend handling that functionality. 🤷‍♀️ I will totally defer to anyone who has a strong opinion about that.

> Not sure what I changed. I added the focus trap similar to popover like we talked about... I will have to investigate and figure out what is going on.

One thing you might want to check is whether there is anything else in the component that might be managing focus, and remove that. E.g. if you included some of the keyboard navigation for the menu, maybe there's a conflict there.

File: packages/patternfly-4/react-core/src/components/ContextSelector/ContextSelector.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      screenReaderLabel,
+      toggleText,
+      searchButtonAriaLabel,
+      searchInputValue,
+      onSearchInputChange,
+      searchInputPlaceholder,
+      ...props
+    } = this.props;
+    return (
+      <div className={css(styles.contextSelector, props.className)} ref={this.parentRef} {...props}>
```
Comment (kmcfaul) [inline packages/patternfly-4/react-core/src/components/ContextSelector/ContextSelector.js]: props.className can just be className.

**Comment by kmcfaul** [comment]
Also noticed what Jen mentioned, it looks like the search is case sensitive at the moment. Additionally, if you backspace out some characters (Aw > A > AW to search AWS, or Opens > Open > OpenS to search OpenShift) to correct the case it doesn't look like it re-searches, you have to back out the entire string. 

When the list is filtered by the text search and you click to select the item, it looks like it is selecting based on the current index for that filtered list, so it selects the wrong item based on the whole refCollection.

File: packages/patternfly-4/react-core/src/components/ContextSelector/ContextSelectorToggle.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        }}
+        className={css(
+          styles.contextSelectorToggle,
+          isFocused && styles.modifiers.focus,
+          isHovered && styles.modifiers.hover,
+          isActive && styles.modifiers.active,
+          className
+        )}
+        type="button"
+        onClick={_event => {
```
Comment (jschuler) [inline packages/patternfly-4/react-core/src/components/ContextSelector/ContextSelectorToggle.js]:
```suggestion
        onClick={() => {
```

File: packages/patternfly-4/react-core/src/components/ContextSelector/ContextSelectorItem.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                styles.contextSelectorMenuListItem,
+                isDisabled && styles.modifiers.disabled,
+                isHovered && styles.modifiers.hover,
+                className
+              )}
+              ref={this.ref}
+              onClick={event => {
+                if (!isDisabled) {
+                  onClick && onClick(event);
+                  onSelect && onSelect(event, children);
```
[the remark this answers is not in the document]
Comment (tlabaj) [reply packages/patternfly-4/react-core/src/components/ContextSelector/ContextSelectorItem.js]: I believe it coms in through the provide/consumer like Select.  I followed that model.

File: packages/patternfly-4/react-core/src/components/ContextSelector/ContextSelectorToggle.js (hunk 2 above)
Comment (tlabaj) [reply packages/patternfly-4/react-core/src/components/ContextSelector/ContextSelectorToggle.js]: I passed the event incase the consumer may want to look at it.

**Comment by christiemolloy** [comment]
@mcarrano should we have an example where you type in the search input, click the search icon and then it searches vs the implementation here where it automatically searches - it just seems as though the button is redundant, and I wanted to get your input on it.

**Comment by jgiardino** [comment]
Great job on this! I just reviewed the updated keyboard and screen reader behavior (in VO). Here's what I noticed:

👍 Behaviors that work as expected:
- Focus lands on the search input when the menu first opens
- For the sighted keyboard-only user, I can only tab through the items in the panel, and focus shifts from the last item to the first item and vice-versa. 
  - Just FYI, I had included a comment in https://github.com/patternfly/patternfly-react/issues/1293 about ways we can refine the keyboard interaction, but my preference would be to defer these as future enhancements
- When the panel is visible, hitting Esc will dismiss the panel

👎 Behaviors that don't work as expected:
- When focus is on the toggle, I should be able to hit Enter to open the panel, but nothing happens.
- When using a screen reader, we should be trapping focus, but currently the screen reader has full access to the DOM. Both the Modal and Popover do this. I think they handle this differently but you might be able to look at one of those components to see how to implement that here. 

And then just a question, not sure who it's for... I noticed that when tabbing through the UI, the Search button can receive focus, but I don't think this button actually does anything. Should we reconsider what type of element this is (for example, is it more of a label for the input than a submit button)? cc'ing @mcarrano on this one. I don't think this is something that needs to be resolved now if we do decide to make changes.

**Comment by mcarrano** [comment]
@christiemolloy @jgiardino I was looking at this earlier today and agree with your comments about the search button being superfluous. Maybe we don't need it. The PF3 design/implementation did not have an explicit search button. @kybaker do you recall the history on this? Is there a reason we included a search button in the design rather than just making this like a type ahead select?

**Comment by jgiardino** [comment]
@mcarrano I just want to add a comment that currently the search icon serves as the visible label for the input, which is required for WCAG AA compliance. It can still be there without being a button.

**Comment by mcarrano** [comment]
Yeah, that makes sense. This was the PF3 design.

Screen Shot 2019-03-18 at 5 01 11 PM

**Comment by jschuler** [comment]
@tlabaj Can you also add the Context Selector to the page demos?

**Comment by tlabaj** [comment]
@mcarrano @jgiardino so will the search button be changing to an icon?

**Comment by jgiardino** [comment]
> Can you also add the Context Selector to the page demos?

👍 I agree, it would be nice to test this in the context of the demo.

> @mcarrano @jgiardino so will the search button be changing to an icon?

I would say that change could be a separate issue, and that we address it in core first. I wouldn't hold this PR up for that since it's not a bug, just a refinement. 

I just reviewed this again. The issue I noted with handling Enter key on the toggle is fixed, which is awesome! 

I tested in JAWS to see if adding `role="dialog" aria-modal="true"` would have any effect on trapping screen reader focus, and it doesn't. 😿 

Considering the differences between popover (which uses tippy) and modal, I wonder if tippy would be our best option for this component. Both require separating the popup element from the `<div>` that contains the main contents, but the modal doesn't require positioning, whereas the popover does. Also, being able to separate the panel and position it on top of the other contents might also let us avoid issues like #1230. I'm fine with capturing this last a11y item in a separate issue if this component is needed and/or this would require a significant amount of work. If we do capture this in a separate issue, then I would remove `aria-modal="true"` since it isn't really modal yet. 

Other than that, there are just a couple of more things I noticed that would be simple updates:

- I just noticed that the order of `id`s on the toggle is reversed. The label id should be first, followed by the button. 
- We can keep `role="dialog"` on the panel, but we should add either `aria-label` or `aria-labelledby` to describe it. 
  - One suggestion is to set `aria-labelledby` to the id of the string defined for `screenReaderLabel`, to avoid adding additional props. But "Selected Project [colon] [dialog]" sounds a little off as the dialog title. Maybe just setting `screenReaderLabel` to "Project" in our example would work for both the toggle and the dialog? This would result in screen readers announcing "Project AWS button collapsed" for the toggle and "Project dialog" for the panel.
  - Another suggestion is to add a prop, similar to `screenReaderLabel` for the toggle, but this prop would be for the panel, and would take a text string and add `aria-label="[text string]"` to the panel. Then in our example, we could define "Select Project" for that prop. This would result in screen readers announcing "Selected Project [colon] AWS button collapsed" for the toggle and "Select Project dialog" for the panel.
  - Let me know if you have a preference on this and I can follow up with design on strings if needed.

**Comment by tlabaj** [comment]
Issue #1615 has been opened and will address remaining accessibility concerns.

**Comment by dlabaj** [comment]
@jgiardino @mcarrano Is enter after typing in search suppose to select the first item in the dropdown from the search results?

**Comment by jschuler** [comment]
IMO button does not make sense if you use typeahead select

**Comment by dlabaj** [comment]
@jschuler @tlabaj I'm going to take this in and if we want to change what happens on enter we can create a separate issue.

**Pull request merged**